Add keyboard shortcuts editor

altran01 opened the pull request:
Keyboard shortcuts for all action types are saved in settings,
they are editable through new dialog available in file menu.
All existing QActions that weren't generated based on some data
were converted to support editing shortcuts for them.

wiktorn commented:
![screenshot from 2017-05-16 14-11-00](https://cloud.githubusercontent.com/assets/3180307/26105257/8edd525a-3a41-11e7-9c53-5d0e5c3687e5.png)

Few things visible  on this screenshot:
* when searching for commands, we can remove groups that do not have matches
* default size of "Add shortcut" dialog could be a bit wider so full message is shown by default ("Press key without m...")
* why grouping items are in grey? They look like inactive

And few not visible on this screenshot:
* It should be "Reset shortcut to default"  not "shortcuts" (it gives intuition, that it might reset all shortcuts)
* when displaying possible conflicts text should be wrapped (maybe each conflict in separate line)

altran01 commented:
About 1st and 3rd comment - QTreeView filtering doesn't support that use-case easily (I added TODO in code and I'll work on it when I fix other issues) - I changed color (only choosing correct one on dark theme) to distinguish empty categories from items. I think we would still like to make categories more noticeable - but I'll use icon for it (for now temporary one from open).
2nd,4th and 5th fixed (for now wrapping - I think it looks better, but we can discuss).

adamiwaniuk commented on src/util/settings/shortcuts.cc:
```diff
@@ -0,0 +1,467 @@
+}
+
+int ShortcutsModel::rowCount(const QModelIndex& parent) const {
+  if (parent.column() > 0)
```
{} or the same line

mkow replied:
It's ok IMO. I don't like the one-line option, but {} are also fine.

adamiwaniuk commented:
When pressiong "Windows" (meta) key 
![windows](https://cloud.githubusercontent.com/assets/21588397/26145161/b729b2d0-3aeb-11e7-880b-855a617983d6.png)

mkow commented:
Another issues:
* We shouldn't allow meta/win shortcuts on Windows, this key is reserved for the OS.
* Numpad keys don't work. On windows they are treated as different keys than their non-numpad counterparts, but we may want to merge them together (because not everyone has numpad).

altran01 commented:
We can discuss the 1st point when you will be in office (or on chat) - we already had a short discussion about it, and I confirmed that you can create shortcuts with win/meta key (if there doesn't exist system shortcut for it - but the same problem exists i.e. for Alt+F4).
I'll look into 2nd point.

mkow commented:
* It's possible to edit shortcut without any selection (enter "aaaaaaaaaa" as a filter and hit [Enter])
* It crashes after adding such shortcut: search for "aaaaaaa", hit [Enter], [A], click "Ok", hit [Esc]
* It's possible to bind numlock, capslock, scroll lock and menu. We should discuss if it's a good idea to support this.
* You can "cheat" the key textbox when editing a shortcut using RMB -> Paste (and paste anything into this field)
* There should be a way to Confirm/Cancel the dialog using keyboard. What about [Ctrl]+[Enter] and [Shift]+[Esc]?

altran01 commented:
Ad 1&2 - forgot to test it after I added option to edit by keyboard shortcut - will fix.
Ad 3 - I don't have strong preference - when I looked at what other programs do in shortcut editors, some allow it and some don't. (BTW - I wasn't able bind scroll lock :P)
Ad 4 - forgot about it (it doesn't break anything since I don't actually read what is written there) - will fix.
Ad 5 - which dialog are you talking about ? In the second dialog for choosing shortcut you can use `Tab` key to change focus, and then you can use buttons and/or Esc&enter/return (I thought it was fine enough). However in the first dialog I indeed by mistake removed ability to save changes using keyboard (you can still cancel using escape) - will fix.

mkow commented on src/util/settings/shortcuts.cc:
```diff
@@ -0,0 +1,473 @@
+  return ptr;
+}
+
+QAction* ShortcutsModel::createQAction(ShortcutType type, QObject* parent, Qt::ShortcutContext context) {
```
wrap

mkow commented on src/util/settings/shortcuts.cc:
```diff
@@ -0,0 +1,473 @@
+  return type_to_shortcut_[type]->createQAction(parent, context);
+}
+
+QAction* ShortcutsModel::createQAction(ShortcutType type, QObject* parent, QIcon icon, Qt::ShortcutContext context) {
```
wrap

mkow commented on src/util/settings/shortcuts.cc:
```diff
@@ -0,0 +1,473 @@
+    return;
+  }
+  if (item->addShortcut(shortcut)) {
+    if (sequence_to_shortcut_.contains(shortcut) && !sequence_to_shortcut_[shortcut].contains(item)) {
```
wrap

mkow commented on src/util/settings/shortcuts.cc:
```diff
@@ -0,0 +1,473 @@
+      }
+      sequence_to_shortcut_[shortcut].append(item);
+    } else {
+      sequence_to_shortcut_[shortcut].append(item);
```
this line is copied in both branches of the if, maybe it'd be better to extract it outside?

altran01 replied:
Right, my mistake.

mkow commented on src/util/settings/shortcuts.cc:
```diff
@@ -0,0 +1,473 @@
+  }
+  if (item->deleteShortcut(shortcut)) {
+    sequence_to_shortcut_[shortcut].removeAll(item);
+    if (sequence_to_shortcut_[shortcut].size() == 1 && !checkIfConflicts(sequence_to_shortcut_[shortcut][0])) {
```
wrap

mkow commented on src/util/settings/shortcuts.cc:
```diff
@@ -0,0 +1,473 @@
+//  addShortcutType(NEW_FILE, global, tr("&New..."), tr("New File"));
+  addShortcutType(SHOW_DATABASE, global, tr("Show database view"));
+  addShortcutType(SHOW_LOG, global, tr("Show log view"));
+  addShortcutType(SHOW_OPTIONS, global, tr("&Options"), tr("Show the Dialog to select applications options"));
```
`dialog`

mkow replied:
Also, these strings seem to occur twice in the code, please extract them to a constants and remove copy-paste ;)

altran01 replied:
In theory they could be different strings (though I used the same :P) - will do.

altran01 replied:
Hmm, where would you want me to put such strings @mkow?

mkow replied:
kk, leave it as is

mkow commented on src/util/settings/shortcuts.cc:
```diff
@@ -0,0 +1,473 @@
+  addShortcutType(SHOW_DATABASE, global, tr("Show database view"));
+  addShortcutType(SHOW_LOG, global, tr("Show log view"));
+  addShortcutType(SHOW_OPTIONS, global, tr("&Options"), tr("Show the Dialog to select applications options"));
+  addShortcutType(SHOW_SHORTCUT_OPTIONS, global, tr("Keyboard shortcuts"), tr("Show the Dialog to customize keyboard shortcuts"));
```
`dialog`

mkow commented on src/util/settings/shortcuts.cc:
```diff
@@ -0,0 +1,473 @@
+  return item;
+}
+
+ShortcutsItem* ShortcutsModel::addShortcutType(ShortcutType type, ShortcutsItem* parent, QString name, QString display_name) {
```
wrap

mkow commented on src/util/settings/shortcuts.cc:
```diff
@@ -0,0 +1,473 @@
+  assert(!type_to_shortcut_.contains(type));
+  type_to_shortcut_[type] = item;
+  auto shortcuts = getShortcuts(type);
+  for (auto& shortcut : shortcuts) {
```
either `const auto&` or `auto`

mkow commented on src/util/settings/shortcuts.cc:
```diff
@@ -0,0 +1,473 @@
+}
+
+bool ShortcutsModel::checkIfConflicts(ShortcutsItem* item) const {
+  for (auto& shortcut : item->shortcuts()) {
```
ditto

mkow commented on src/util/settings/shortcuts.cc:
```diff
@@ -0,0 +1,473 @@
+
+void ShortcutsModel::resetShortcutsForType(ShortcutType type) {
+  auto item = type_to_shortcut_[type];
+  for (auto shortcut : item->shortcuts()) {
```
const ref

mkow commented on src/util/settings/shortcuts.cc:
```diff
@@ -0,0 +1,473 @@
+  for (auto shortcut : item->shortcuts()) {
+    removeShortcut(type, shortcut);
+  }
+  for (auto shortcut : getShortcuts(type)) {
```
const ref

mkow commented on src/util/settings/shortcuts.cc:
```diff
@@ -0,0 +1,473 @@
+void ShortcutsModel::resetShortcutsToDefault(ShortcutType type) {
+  auto shortcuts = defaultShortcuts();
+  auto item = type_to_shortcut_[type];
+  for (auto shortcut : item->shortcuts()) {
```
const ref

mkow commented on src/util/settings/shortcuts.cc:
```diff
@@ -0,0 +1,473 @@
+  for (auto shortcut : item->shortcuts()) {
+    removeShortcut(type, shortcut);
+  }
+  for (auto shortcut : shortcuts.value(type)) {
```
const ref

mkow commented on src/util/settings/shortcuts.cc:
```diff
@@ -0,0 +1,473 @@
+}
+
+void ShortcutsModel::resetAllShortcutsToDefaults() {
+  for (auto iter : type_to_shortcut_) {
```
const ref ?

mkow commented on src/visualization/panel.cc:
```diff
@@ -315,27 +319,33 @@ void VisualizationPanel::initOptionsPanel() {
-      ":/images/digram_icon.png", icon_color), tr("&Digram"), this);
+  digram_action_ = util::settings::shortcuts::ShortcutsModel::getShortcutsModel()->createQAction(
+        util::settings::shortcuts::VISUALIZATION_DIGRAM, this,
+        util::getColoredIcon(":/images/digram_icon.png", icon_color), Qt::WidgetWithChildrenShortcut);
```
wrap

mkow commented on src/visualization/panel.cc:
```diff
@@ -315,27 +319,33 @@ void VisualizationPanel::initOptionsPanel() {
-      ":/images/trigram_icon.png", icon_color), tr("&Trigram"), this);
+  trigram_action_ = util::settings::shortcuts::ShortcutsModel::getShortcutsModel()->createQAction(
+        util::settings::shortcuts::VISUALIZATION_TRIGRAM, this,
+        util::getColoredIcon(":/images/trigram_icon.png", icon_color), Qt::WidgetWithChildrenShortcut);
```
wrap

mkow commented on src/visualization/panel.cc:
```diff
@@ -362,10 +372,13 @@ void VisualizationPanel::initOptionsPanel() {
-  QAction* show_more_options_action = new QAction(QIcon(":/images/more.png"),
-      tr("More options"), this);
+
+  QAction* show_more_options_action = util::settings::shortcuts::ShortcutsModel::getShortcutsModel()->createQAction(
```
wrap

mkow commented on src/visualization/trigram.cc:
```diff
@@ -133,23 +133,20 @@ void TrigramWidget::prepareOptions(QMainWindow* visualization_window) {
-  cube_action_->setIcon(
-      util::getColoredIcon(":/images/cube.png", icon_color, false));
+  cube_action_ = util::settings::shortcuts::ShortcutsModel::getShortcutsModel()->createQAction(
+        util::settings::shortcuts::TRIGRAM_CUBE, this, util::getColoredIcon(":/images/cube.png", icon_color, false));
```
wrap

mkow commented on src/visualization/trigram.cc:
```diff
@@ -133,23 +133,20 @@ void TrigramWidget::prepareOptions(QMainWindow* visualization_window) {
-  cylinder_action_->setIcon(
-      util::getColoredIcon(":/images/cylinder.png", icon_color, false));
+  cylinder_action_ = util::settings::shortcuts::ShortcutsModel::getShortcutsModel()->createQAction(
+        util::settings::shortcuts::TRIGRAM_CYLINDER, this, util::getColoredIcon(":/images/cylinder.png", icon_color, false));
```
wrap

mkow commented on src/visualization/trigram.cc:
```diff
@@ -133,23 +133,20 @@ void TrigramWidget::prepareOptions(QMainWindow* visualization_window) {
-  sphere_action_->setIcon(
-      util::getColoredIcon(":/images/sphere.png", icon_color));
+  sphere_action_ = util::settings::shortcuts::ShortcutsModel::getShortcutsModel()->createQAction(
+        util::settings::shortcuts::TRIGRAM_SPHERE, this, util::getColoredIcon(":/images/sphere.png", icon_color));
```
wrap

mkow commented:
> Ad 5 - which dialog are you talking about ? In the second dialog for choosing shortcut you can use Tab key to change focus, and then you can use buttons and/or Esc&enter/return (I thought it was fine enough). However in the first dialog I indeed by mistake removed ability to save changes using keyboard (you can still cancel using escape) - fixed.

That jumping with <kbd>Tab</kbd> to "Ok" button isn't too convenient and <kbd>Tab</kbd> and then <kbd>Enter</kbd>/<kbd>Esc</kbd> isn't too intuitive IMO. <kbd>Ctrl</kbd>+<kbd>Enter</kbd> is IMO a standard way for accepting dialogs/forms that capture plain <kbd>Enter</kbd>.

altran01 commented on include/util/settings/shortcuts.h:
```diff
@@ -0,0 +1,162 @@
+  void addShortcut(util::settings::shortcuts::ShortcutType type, QKeySequence& shortcut);
+  void removeShortcut(util::settings::shortcuts::ShortcutType type, const QKeySequence& shortcut);
+  ShortcutsItem* itemFromIndex(const QModelIndex& index) const;
+  QModelIndex indexFromItem(ShortcutsItem* item) const;
```
Due to how Qt interfaces are defined I would need to use const_cast somewhere if I add const here - I'm leaning towards leaving it as it is - how do you feel about it?

mkow replied:
Ok, then leave it as is.

altran01 commented on src/ui/shortcutedit.cc:
```diff
@@ -0,0 +1,54 @@
+  clear();
+}
+
+void ShortcutEdit::keyPressEvent(QKeyEvent* event) {
```
This is part of Qt interface (in .h file you can see override).

mkow replied:
ack

altran01 commented on src/ui/shortcutedit.cc:
```diff
@@ -0,0 +1,54 @@
+    // and handle it through checkboxes
+    return;
+  }
+  if (key == Qt::Key_Shift || key == Qt::Key_Control || key == Qt::Key_Meta || key == Qt::Key_Alt) {
```
I wasn't able to break current checks on neither platforms I tested it on (including macOS) - AltGr always comes with modifiers, but I can include it if you will feel safer with it :)

mkow replied:
Same for me, but let's stay on the safe side and check it, as the specification doesn't guarantee that it must come with other modifiers ;)

altran01 commented on src/ui/shortcutssettings.cc:
```diff
@@ -0,0 +1,230 @@
+}
+
+QKeySequence ShortcutEditDialog::getSequence() {
+  int key = ui->lineEdit->key();
```
I would agree if not for the fact that Qt itself uses ints for it - so if it will bite us we won't be able to do anything about it either way (`key()` returns int) :P

mkow replied:
ack

altran01 commented on src/ui/veles_mainwindow.cc:
```diff
@@ -154,56 +155,70 @@ void VelesMainWindow::init() {
+  shortcuts_act_ = util::settings::shortcuts::ShortcutsModel::getShortcutsModel()->createQAction(
+        util::settings::shortcuts::SHOW_SHORTCUT_OPTIONS, this, Qt::ApplicationShortcut);
+  shortcuts_act_->setStatusTip(
+      tr("Show the Dialog to customize keyboard shortcuts"));
```
Its a copy paste from other dialog status tip :P

altran01 commented:
> That jumping with Tab to "Ok" button isn't too convenient and Tab and then Enter/Esc isn't too intuitive IMO. Ctrl+Enter is IMO a standard way for accepting dialogs/forms that capture plain Enter.

Done.

mkow commented on CMakeLists.txt:
```diff
@@ -310,17 +310,17 @@ target_link_libraries(unpyc veles_db parser veles_network)
+  ${SRC_DIR}/ui/optionsdialog.ui
+  ${SRC_DIR}/ui/searchdialog.ui
+  ${SRC_DIR}/ui/shortcutssettings.ui
+  ${SRC_DIR}/ui/shortcutselection.ui
```
shortcutselection < shortcutssettings
